### PR TITLE
Prevent possible warning in `COTMigrationUtil::get_post_or_object_meta()`

### LINKED ISSUE #35543: Calling `woocommerce_wp_select` in WooCommerce > Settings context throws a notice
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Plugins use the `woocommerce_wp_select` function (and the rest `woocommerce_wp_XXXX` functions) to display elements under WooCommerce > Settings.

From [this commit](https://github.com/woocommerce/woocommerce/commit/b4fcbd80a411d0e428df2dacc46b6187f3b6fe72) onwards, these functions use the global `$post` object, which is null in this context -- before these changes, the global `$thepostid` was used: [https://github.com/woocommerce/woocommerce/commit/b4fcbd80a411d0e428df2dacc46b6187f3b6fe72#diff-9e238275778983fe5abc7c[…]c28ede7a39c6633ee91d226L192](https://github.com/woocommerce/woocommerce/commit/b4fcbd80a411d0e428df2dacc46b6187f3b6fe72#diff-9e238275778983fe5abc7c4431df963b9327256a9c28ede7a39c6633ee91d226L192)

This makes admin pages show up like this:  because WooCommerce tries to access the ID property of a NULL object.

I was able to replicate this issue by activating All Products for WooCommerce Subscriptions and navigating to the WooCommerce > Settings > Subscriptions tab. This is the error that shows up there: 

```
[10-Nov-2022 12:57:05 UTC] PHP Notice:  Trying to get property 'ID' of non-object in /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/src/Internal/Utilities/COTMigrationUtil.php on line 99
[10-Nov-2022 12:57:05 UTC] PHP Stack trace:
[10-Nov-2022 12:57:05 UTC] PHP   1. {main}() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-admin/admin.php:0
[10-Nov-2022 12:57:05 UTC] PHP   2. do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-admin/admin.php:259
[10-Nov-2022 12:57:05 UTC] PHP   3. WP_Hook->do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/plugin.php:517
[10-Nov-2022 12:57:05 UTC] PHP   4. WP_Hook->apply_filters() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:332
[10-Nov-2022 12:57:05 UTC] PHP   5. WC_Admin_Menus->settings_page() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:308
[10-Nov-2022 12:57:05 UTC] PHP   6. WC_Admin_Settings::output() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-menus.php:286
[10-Nov-2022 12:57:05 UTC] PHP   7. include() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-settings.php:154
[10-Nov-2022 12:57:05 UTC] PHP   8. do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/views/html-admin-settings.php:40
[10-Nov-2022 12:57:05 UTC] PHP   9. WP_Hook->do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/plugin.php:517
[10-Nov-2022 12:57:05 UTC] PHP  10. WP_Hook->apply_filters() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:332
[10-Nov-2022 12:57:05 UTC] PHP  11. WC_Subscriptions_Admin::subscription_settings_page() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:308
[10-Nov-2022 12:57:05 UTC] PHP  12. woocommerce_admin_fields() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/admin/class-wc-subscriptions-admin.php:1136
[10-Nov-2022 12:57:05 UTC] PHP  13. WC_Admin_Settings::output_fields() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/wc-admin-functions.php:171
[10-Nov-2022 12:57:05 UTC] PHP  14. do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/class-wc-admin-settings.php:737
[10-Nov-2022 12:57:05 UTC] PHP  15. WP_Hook->do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/plugin.php:517
[10-Nov-2022 12:57:05 UTC] PHP  16. WP_Hook->apply_filters() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:332
[10-Nov-2022 12:57:05 UTC] PHP  17. WCS_ATT_Admin::subscription_schemes_content() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:308
[10-Nov-2022 12:57:05 UTC] PHP  18. do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce-all-products-for-subscriptions/includes/admin/class-wcs-att-admin.php:250
[10-Nov-2022 12:57:05 UTC] PHP  19. WP_Hook->do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/plugin.php:517
[10-Nov-2022 12:57:05 UTC] PHP  20. WP_Hook->apply_filters() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:332
[10-Nov-2022 12:57:05 UTC] PHP  21. WCS_ATT_Meta_Box_Product_Data::subscription_scheme() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:308
[10-Nov-2022 12:57:05 UTC] PHP  22. include() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce-all-products-for-subscriptions/includes/admin/meta-boxes/class-wcs-att-meta-box-product-data.php:233
[10-Nov-2022 12:57:05 UTC] PHP  23. do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce-all-products-for-subscriptions/includes/admin/meta-boxes/views/subscription-scheme.php:22
[10-Nov-2022 12:57:05 UTC] PHP  24. WP_Hook->do_action() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/plugin.php:517
[10-Nov-2022 12:57:05 UTC] PHP  25. WP_Hook->apply_filters() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:332
[10-Nov-2022 12:57:05 UTC] PHP  26. WCS_ATT_Meta_Box_Product_Data::subscription_scheme_content() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-includes/class-wp-hook.php:308
[10-Nov-2022 12:57:05 UTC] PHP  27. woocommerce_wp_select() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce-all-products-for-subscriptions/includes/admin/meta-boxes/class-wcs-att-meta-box-product-data.php:287
[10-Nov-2022 12:57:05 UTC] PHP  28. Automattic\WooCommerce\Utilities\OrderUtil::get_post_or_object_meta() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/includes/admin/wc-meta-box-functions.php:201
[10-Nov-2022 12:57:05 UTC] PHP  29. Automattic\WooCommerce\Internal\Utilities\COTMigrationUtil->get_post_or_object_meta() /Users/dimitrioskytros/Local Sites/wp/app/public/wp-content/plugins/woocommerce/src/Utilities/OrderUtil.php:59
```

This is not an extension-specific error though. It can be triggered by calling `woocommerce_wp_select` in the WooCommerce > Settings context manually.

### Expected behavior

The expected behavior is that plugins should be able to call `woocommerce_wp_select` in the WooCommerce > Settings context to display select elements.

### Actual behavior

The actual behavior is that when plugins call `woocommerce_wp_select` in the WooCommerce > Settings context, a notice shows up.

### Steps to reproduce

Write a snippet to call `woocommerce_wp_select` in a WooCommerce > Settings hook or enable All Products for WooCommerce Subscriptions and navigate to WooCommerce > Settings > Subscriptions.

### WordPress Environment

Fresh WordPress install with the latest WooCommerce version active.

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by jorgeatorres**
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR makes sure that `$post->ID` is set before attempting to use it inside calls to [`COTMigrationUtil::get_post_or_object_meta()`](https://github.com/woocommerce/woocommerce/blob/77c87bc51611e3ed6e4062187e43d1b9124278ff/plugins/woocommerce/src/Internal/Utilities/COTMigrationUtil.php#L92).

This function mostly exists for HPOS/posts compat reasons and serves functions `woocommerce_wp_select()`, `woocommerce_wp_text_input()`, `woocommerce_wp_hidden_input()`, etc. in [`wc-meta-box-functions.php`](https://github.com/woocommerce/woocommerce/blob/b4fcbd80a411d0e428df2dacc46b6187f3b6fe72/plugins/woocommerce/includes/admin/wc-meta-box-functions.php).

Closes #35543.

### How to test the changes in this Pull Request:

- [X] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

Per #35543, you should "write a snippet to call `woocommerce_wp_select` in a WooCommerce > Settings hook […]".

The issue can also be replicated by just calling those functions in almost any context where the `$post` global is not set, but to replicate the exact conditions of #35543, the following can be used:

1. Create a test plugin/snippet with the following code (or add it to the theme's `functions.php`)
   ```php
   add_action(
	   'woocommerce_settings_general',
	   function() {
		   woocommerce_wp_select(
			   array(
				   'id'          => 'my-favorite-topping',
				   'value'       => 'pepperoni',
				   'label'       => 'Pizza Topping',
				   'options'     => [
					   'pineapple' => 'Pineapple',
					   'salami'    => 'Salami',
					   'pepperoni' => 'Pepperoni',
					   'onion'     => 'Onion',
				   ],
				   'desc_tip'    => true,
				   'description' => 'Toppings are a serious matter. Choose wisely.',
			   )
		   );
	   },
	   999
   );
   ```
2. Go to WC > Settings and scroll to the bottom of the page.
3. You should see a dropdown named "Pizza Topping" (no need to make a selection, though the best choice has been pre-selected for you).
4. Now, to the "errors" part:
   - On `trunk`, you should see a warning in the log such as `PHP Notice:  Trying to get property 'ID' of non-object in […]`.
   - On this branch, no warning should be logged.

<!-- End testing instructions —>

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [X] Have you included testing instructions?

<!-- Mark completed items with an [x] —>

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Utilities/COTMigrationUtil.php [hunk 1]
```diff
@@ -96,7 +96,7 @@ public function get_post_or_object_meta( ?WP_Post $post, ?\WC_Data $data, string
 			}
 			return $data->get_meta( $key, $single );
 		} else {
-			return get_post_meta( $post->ID, $key, $single );
+			return isset( $post->ID ) ? get_post_meta( $post->ID, $key, $single ) : false;
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Utilities/COTMigrationUtil.php:99]: Just noting that I've decided to use `false` as "default" value when `$post->ID` is not set, just to make `get_post_or_object_meta ()` mimic what `get_post_meta()` does: return `false` for invalid `$post_id` ([see codex](http://developer.wordpress.org/reference/functions/get_post_meta/)). This, however, makes probably little difference in real use.

**Comment by PanosSynetos** [comment]
Hey @jorgeatorres , thanks for working on this! - Once this is ready to merge, could you please share a zip so we can test it on our environments? cc @jimjasson

**Comment by rrennick** [comment]
@jorgeatorres There were issues in the PHP 7.4 unit tests.

**Comment by jorgeatorres** [comment]
> @jorgeatorres There were issues in the PHP 7.4 unit tests.

Thanks @rrennick. Don't know why those failed but after re-running the job, they are now passing.

**Review by rrennick — APPROVED** [review]
LGTM. Tested great.

**Comment by rrennick** [comment]
> Once this is ready to merge, could you please share a zip so we can test it on our environments?

@PanosSynetos I went ahead and merged this as it was a single line change. If you want to test this you can use `pnpm run build:zip` in the `plugins/woocommerce` folder to create a zip.

**Comment by PanosSynetos** [comment]
Thanks, we'll test it out! (sorry I missed the ping)

**Comment by PanosSynetos** [comment]
Tested, we're good 👌 and the warning is gone

**Pull request merged**